### ISSUE TITLE
Support for Open API spec 3.0

### ISSUE THREAD

**Issue by ifraixedes** [go-swagger/go-swagger#1122]
Only to bear in mind, is there any plan to release and update to support Open API spec 3.0 https://swagger.io/announcing-openapi-3-0/

Thanks!

**Comment by seifer** [comment]
Very interesting!!

**Comment by casualjim** [comment]
for definitions 2.0 and 3.0 aren't very different because I wanted a full json schema implementation to work instead of the subset that swagger (openapi) 2.0 prescribed
in 3.0 they expanded it to properties that are already supported in our validator
most of the changes in 3.0 pertain to how operations are described with their parameters and responses
that has a big impact on the code generator but if I'm guessing nothing in there is a combinatorial nightmare unlike json schema is
I'll start to work on a new spec implementation in go-openapi/spec3 because I do want to correct some thngs like not using golang's map as a map but rather use sorted maps

**Comment by casualjim** [comment]
spec will be here: https://github.com/go-openapi/spec3

**Comment by tamalsaha** [comment]
@casualjim, is there any update on openapi 3 support?

**Comment by JunliWang** [comment]
@casualjim, check-in again, is there a timeline for openapi 3 support? Or early adoption on another branch?

**Comment by fredbi** [comment]
@JunliWang hello.
We were just discussing that yesterday. It's really a ton of work to lift it to v3.
I am not sure there a many volunteers out there to produce such an effort.

Personnally, I would concentrate my efforts on getting the product stable with full V2 support and more than good enough. It is already good enough for most use cases, and probably at or close to top of this class of products.

**Comment by SandyWalsh** [comment]
@fredbi the problem is, as long as the spec reads `swagger: '2.0'` we will have issues using the spec for other tools, doc generators, etc.

**Comment by casualjim** [comment]
there are a lot more issues with 3.0. it has expressions, a vastly different syntax and so on.
Somebody else is more than welcome to step up but I don't have time to implement and mantain such a large specification as main contributor.

**Comment by fredbi** [comment]
@casualjim Me neither. A "Swager 2.1" fixing some inconsistencies in 2.0 would be fair enough for me.
@SandyWalsh : we are basically saying the same thing  .It would be rathr easy to be more tolerant with the version announced by the spec, while continuing to check it against 2.0.

**Comment by SandyWalsh** [comment]
Thanks @casualjim @fredbi ... totally understand. Perhaps if I can get some cycles I'll try to contribute. I think I have some tricks I can do to keep on 2.0. Keep up the great work!

**Comment by theherk** [comment]
I'm curious the status of this. `validate` still fails, but `serve` seems to work just fine.

For anybody else curious, I still maintain the v3 definition, I just use api-spec-converter to autogen the swagger v2 definition.

**Comment by msolimans** [comment]
Same here, it fails when I tried to `validate`, `serve` works but not perfectly when using `oneOf` in the response it gave me no response examples.

Any updates!

**Comment by casualjim** [comment]
I don't have time to work on an openapi 3.0 specification.

**Comment by fenollp** [comment]
For validation, Google’s `gnostic` generates/maintains a schema and some tooling. Hopefully this helps :)

**Comment by kappj** [comment]
I wanna bump this as all the other tooling I use with my OAS supports V3 now, and this project is the only thing that's keeping me from migrating to OAS 3.

**Comment by Yaima** [comment]
It looks like OpenAPI 3.0 supports multiple hosts. What's the timeline for this release? We need to be able to specify multiple hosts: development, QA, and production.

**Comment by fredbi** [comment]
@Yaima for such a simple use case, such a breaking change is perhaps not needed: 
when it comes to generate a server for instance, it suffices to publish a slightly modified spec (e.g. override host with `swagger mixin`). For generated clients, the host in the spec is just a default which is easily overriden in your client's code.

**Comment by Yaima** [comment]
How does  [mixin-OPTIONS] look like? Could this only have the host information that I need?

**Comment by casualjim** [comment]
yes it's a deep merge of a spec

**Comment by Yaima** [comment]
@casualjim, will something like this work? `swagger swagger.json mixin swagger-prod.json` What does [OPTIONS] refer to? Is it a json, yml? `swagger [OPTIONS] mixin [mixin-OPTIONS]`

**Comment by casualjim** [comment]
like this: `swagger mixin --output ./swagger-final.json ./swagger.json ./swagger-prod.json`

**Comment by Yaima** [comment]
I am getting this error: 
`2018/11/27 14:36:08 args[0] = ./swagger.json
2018/11/27 14:36:08 args[1:] = [./swagger-prod.json]
unexpected end of JSON input` Any idea why? Could be the content of the .swagger-prod? I am trying to keep only the fields to change. ie: 
```
{
  "host": "newhost.com",
  "securityDefinitions": {
    "oauth2": {
      "type": "oauth2",
      "flow": "application",
      "tokenUrl": "https://newhost.com/v1/oauth/token"
    }
}
```

**Comment by Yaima** [comment]
@casualjim what does this mean?
2018/11/29 10:32:39 args[0] = ./swagger.json
2018/11/29 10:32:39 args[1:] = [./swagger-prod.json]
2018/11/29 10:32:39 SecurityDefinitions entry 'oauth2' already exists in primary or higher priority mixin, skipping

This is the information I am changing, but the final json it's not being affected by them.

```
{
  "host": "abc.def",
  "securityDefinitions": {
    "oauth2": {
      "type": "oauth2",
      "flow": "application",
      "tokenUrl": "https://abc.def/v1/oauth/token"
    }
  }
}
```

**Comment by Yaima** [comment]
@casualjim If I switch the order of the arguments, it seems to work. Is this correct?

```
swagger mixin -o ./swagger-final.json ./swagger-prod.json ./swagger.json
2018/11/29 10:40:44 args[0] = ./swagger-prod.json
2018/11/29 10:40:44 args[1:] = [./swagger.json]
2018/11/29 10:40:44 SecurityDefinitions entry 'oauth2' already exists in primary or higher priority mixin, skipping
```

**Comment by fredbi** [comment]
@Yaima could you please move your questions to a dedicated issue? that would be nicer for people wanting to continue the OpenAPI3 discussion. Thank you.

**Comment by Ark-kun** [comment]
Does go-swagger support JsonSchema-based OpenAPI specs that use oneOf?

**Comment by philsturgeon** [comment]
oneOf is part of OpenAPI v3.0, it does not require JSON Schema-proper in order to do that, just full OpenAPI v3.0.

**Comment by beoran** [comment]
How about making use of the work done by this project? https://github.com/getkin/kin-openapi

**Comment by recrack** [comment]
How is in progress?

**Comment by leslie-wang** [comment]
I'm trying to use swagger to document Openstack Ironic Project. One of the APIs may return different body depending on request query string, for example, if query string has `detail=true`, it returns `node_detail` payload, otherwise, return `node_overview` payload. Thus, `oneOf` is the solution to address it. At the same time, I do want to use go-swagger to generate both client and server packages. So I try to add notes here.

**Comment by casualjim** [comment]
you can put the query string in the path and that way you can achieve this result with openapi 2.0

**Comment by leslie-wang** [comment]
Unfortunately, Openstack's APIs have already been defined, and can not changed by me. I'm just trying to use swagger format to document it, and potentially use go-swagger to generate codes with golang.

**Comment by kappj** [comment]
@leslie-wang I gave up waiting for this and customised my own openapi-generator earlier this year so I could start using OASv3.

**Comment by thekrishnavamsi** [comment]
Any updates or ETA? Even first version without oneof and all will be great.

**Comment by chriswalz** [comment]
Any updates on this?

**Comment by donggangcj** [comment]
I convert the OAS3.0 to OAS2.0 by some tools.

**Comment by aeneasr** [comment]
With OAS3, it's now possible to express [Cookie Auth](https://swagger.io/docs/specification/authentication/cookie-authentication/) which doesn't work with OAS3 as `securityDefinitions.#.in` may only be one of `header` and `query`.

Is there a general willingness to move this to OAS3 or will go-swagger simply remain OAS2 only?

**Comment by aaahrens** [comment]
I would love this for cookie auth + different url loading.

**Comment by applekate** [comment]
FYI. I've been using OpenAPI Generator to generate Go clients and server-side code and I'm happy with the result so far.

Ref: https://github.com/OpenAPITools/openapi-generator

**Comment by qiangli** [comment]
"$ref:" is an OpenAPI 3.0 which obviously go-swagger does not support. https://swagger.io/docs/specification/using-ref/

Removing $ref certainly will make go-swagger happy but we will lose the ability to test our apis with other standard tools. (I'm using Swagger UI)

Go-swagger author was reluctant to move to 3.0 since at least 2017 https://github.com/go-swagger/go-swagger/issues/1122 in spite of strong demands.

Should we reevaluate and move to other tools like others? I  really value Swagger and consider it a valuable tool saving us dev/test time.

**Comment by nicesj** [comment]
> FYI. I've been using OpenAPI Generator to generate Go clients and server-side code and I'm happy with the result so far.
> 
> Ref: https://github.com/OpenAPITools/openapi-generator

Hello, I also visited that site as following your comments.
And I successfully generated a server stub code using the specification file.
However, I could not find how it validates the request and the response object by the specification in runtime like the code that is generated by the go-swagger.

Could you share how did you validate the objects response/request by the specification?

**Comment by tboerger** [comment]
Has anybody found alternatives to go-swagger for generating the server? I really dislike the templates of the openapi generator. The project https://github.com/deepmap/oapi-codegen is generating useable code based on an interface, but it doesn't provide any callback for authentication like go-swagger does.

**Comment by philsturgeon** [comment]
Nobody ever likes the provided templates for any codegenerators. I don't know if thats a problem with the code generators templates, or if its programmers nature, but you can make your own templates... Take the go-swagger output and make that into a template? Solve it for everyone?

**Comment by tboerger** [comment]
But the official code generator even dictates the folder and file name structure, that can't be changed.

**Comment by philsturgeon** [comment]
That's not true. Please dig into the documentation and ask their community for help if you get stuck.

**Comment by tboerger** [comment]
Ok, let me rephrase that, it's not possible to really customize it without the need to compile the generator on my own because it got all the paths statically defined within the java class. You are only able to customize the content of the predefined templates which is not defined like I would expect from properly defined packages in Go.

**Comment by philsturgeon** [comment]
This is not a support channel for openapi-generator and its not the right place for feedback, so lets not hijack the thread.

**Comment by Sufian22** [comment]
> "$ref:" is an OpenAPI 3.0 feature which obviously go-swagger/validate does not support. https://swagger.io/docs/specification/using-ref/
> 
I could be able to use go-swagger to generate code from 2.0 specification using **$ref**'s, I don't know where you get that it's an OpenAPI 3.0 feature.. https://swagger.io/docs/specification/2-0/basic-structure/

If the OpenAPI 3.0 is correctly specified and without explicit 3.0 syntax/features (e.g anyOf, oneOf, Cookies Auth..), it's easy to do a conversion to OpenAPI 2.0 and use this tool.

**Comment by fredbi** [comment]
This threads starts looking like a ranting/trolling alley, which is not the objective of this issue.
I am contemplating locking this thread.  Any objections?

**Comment by erihanse** [comment]
@fredbi 
How can we stay updated on the issue?

**Comment by casualjim** [comment]
There is no point in continuing to ask. I don't have the time to spend on a openapi 3.0 implementation. The spec is way more complex and this requires more work than a single person can conceivably handle. 
This is a 2-3 person hobby project with many one-off small bug fixes at this point. Building a community and getting people to be more actively involved requires a significant investment of time to do the conferences and so forth. There are people who like doing that but I'm not one of them.

**Comment by fredbi** [comment]
I concur with @casualjim, it is time to lock that thread. The current maintainers are not going to make it to OAS3 ever. If some people are interested, PR's are welcome, but that's indeed a lot of work.
